client/blb/client.go

```diff
@@ -1127,109 +1126,123 @@ func (cli *Client) readOneTractReplicated(
+				ch <- tractResultRepl{
+					tractResult: tractResult{0, 0, err, badVersionHost},
+					thisB:       thisBLoc,
+				}
 				return
 			}
 
-			// TODO(eric): fill in otherHosts when ts-ts cancellation is done.
-			read, err := cli.tractservers.ReadInto(ctx, host, nil, reqID, tract.Tract, tract.Version, thisB, thisOffset)
-			if err == core.ErrVersionMismatch {
-				badVersionHost = host
-			}
-			if err != core.NoError && err != core.ErrEOF {
-				// TODO(eric): redo bad TS reporting mechanism.
-				ch <- tractResult{0, 0, err, badVersionHost}
-				return
-			}
-			log.V(1).Infof("read %s from tractserver at address %s: %s", tract.Tract, host, err)
-			// In case of short read, i.e., read < len(thisB), we should
-			// pad the result with 0's. This doesn't need to be done
-			// explicitly if 'thisB' given to us starts to be empty.
-			// However, just in case it's not, we will ensure that by
-			// overwritting untouched bytes with 0's.
-			for i := read; i < len(thisB); i++ {
-				thisB[i] = 0
+			result := readFunc(host, n, thisBLoc)
+
+			ch <- tractResultRepl{
+				tractResult: result,
+				badReadHost: host,
+				thisB:       thisBLoc,
 			}
-			ch <- tractResult{len(thisB), read, err, badVersionHost}
-			return
 		}
 
-		// Backup request logic involves the following:
-		// 1) Send the first read to the first host (item 0 available from the orderCh) in a goroutine.
-		// 2) Start another goroutine to issue backup requests. Each subsequent request is spawned after the
-		// configured backup time.After delay time expires.
-		// 3) Block until the first read is returned from the goroutines servicing the reads. Note that if the
-		// the request returned fails, we accept that and continume on to the fallback phase for sequential reads.
-		// We could optimize this later and scan for the first successful backup request, but that may not be
-		// worth the latency savings.
+		var wg sync.WaitGroup
+		nb := min(maxNumBackups+1, len(order))
+		wg.Add(nb)
 		go func() {
-			nb := min(maxNumBackups+1, len(order))
-			for i := 0; i < nb; i++ {
+			for n := 0; n < nb; n++ {
 				select {
-				case <-cli.backupReadState.backupDelayFunc(time.Duration(i) * delay):
-					go readFunc(order[i])
+				case <-cli.backupReadState.delayFunc(time.Duration(n) * delay):
+					thisBLoc := make([]byte, len(thisB))
+					go readFuncBackup(&wg, n, thisBLoc)
 				case <-ctx.Done():
-					log.V(1).Infof("backup read %s-%d to tractserver cancelled", tract.Tract, i)
+					log.V(1).Infof("backup read %s-%d to tractserver cancelled", tract.Tract, n)
 					break
 				}
 			}
 		}()
+		go func() {
+			wg.Wait()
+			close(ch)
+		}()
 
-		// take the first result.
-		*result = <-ch
-		if result.err == core.NoError {
-			return
+		// Take the first result or mark bad hosts from being read during the
+		// fallback phase below.
+		for res := range ch {
+			if res.err == core.NoError {
+				*result = res.tractResult
+				copy(thisB, res.thisB)
+				return
```

Comment:
I still think returning without draining ch has the potential to leak goroutines: what if two of them both try to send on the channel at right about the same time (so it's too late for cancellation to do anything)?

I think always reading a fixed number of things (before returning) will solve this and make the logic cleaner. You can just cancel() after you get a success so that the other ones return immediately.